### PR TITLE
Improve deduction guide for ExperimentalHyperGeometry::Point

### PR DESCRIPTION
**Pull request by masterleinad**
In response to https://github.com/arborx/ArborX/pull/916#issuecomment-1662433663 and https://github.com/arborx/ArborX/pull/916#discussion_r1282012041.
- ~~we ensure that the template type argument is a floating point type.~~
- the deduction guide now works for all dimensions, and uses `double` if any of the arguments is a `double` and `float` otherwise.

### CODE REVIEW

File: src/geometry/ArborX_HyperPoint.hpp [hunk 1]
```diff
@@ -22,6 +22,8 @@ namespace ArborX::ExperimentalHyperGeometry
 template <int DIM, class FloatingPoint = float>
 struct Point
 {
+  static_assert(std::is_floating_point_v<FloatingPoint>);
```
Comment (aprokop) [inline src/geometry/ArborX_HyperPoint.hpp:25]: I don't think we want this. It should still be possible for a user to build `Point<D, int>` if desired.

File: src/geometry/ArborX_HyperPoint.hpp [hunk 2]
```diff
@@ -34,15 +36,11 @@ struct Point
...
-// Deduction guides
-template <class T>
-Point(T x) -> Point<1, T>;
-
-template <class T>
-Point(T x, T y) -> Point<2, T>;
-
-template <class T>
-Point(T x, T y, T z) -> Point<3, T>;
+template <typename... T>
```
Comment (aprokop) [inline src/geometry/ArborX_HyperPoint.hpp:39]:
```suggestion
// Deduction guide
template <typename... T>
```

File: src/geometry/ArborX_HyperPoint.hpp [hunk 3]
```diff
@@ -34,15 +36,11 @@ struct Point
...
+Point(T...)
+    -> Point<sizeof...(T), std::conditional_t<
+                               (... || std::is_same_v<std::decay_t<T>, double>),
```
Comment (aprokop) [inline src/geometry/ArborX_HyperPoint.hpp:42]: How does this work? Which argument will be decayed? First one, or any of them? If we do `(0, 0.2)`, would that still decay to `double`?

File: src/geometry/ArborX_HyperPoint.hpp (hunk 1 above)
Comment (masterleinad) [reply src/geometry/ArborX_HyperPoint.hpp:25]: Then, we shouldn't call the template argument `FloatingPoint`.

File: src/geometry/ArborX_HyperPoint.hpp (hunk 3 above)
Comment (masterleinad) [reply src/geometry/ArborX_HyperPoint.hpp:42]: All of them are cast to `double` if any of the arguments is a double and if not all of them are cast to `float`.
`(0, 0.2)` would give a `Point<2, double>`.

Comment (masterleinad) [reply src/geometry/ArborX_HyperPoint.hpp:42]: The question still is how to get around `Wc++11-narrowing`-type warnings.

File: src/geometry/ArborX_HyperPoint.hpp (hunk 1 above)
Comment (aprokop) [reply src/geometry/ArborX_HyperPoint.hpp:25]: I agree in principle. Not changing as part of this PR, though.

**Comment by aprokop** [comment]
What happens if you do `Point()`? Parameter packs can be empty, right? Would it create `Point<0, float>`?

**Comment by masterleinad** [comment]
> What happens if you do `Point()`? Parameter packs can be empty, right? Would it create `Point<0, float>`?

Yes, you get `Point<0, float>`.

File: src/geometry/ArborX_HyperPoint.hpp (hunk 3 above)
Comment (masterleinad) [reply src/geometry/ArborX_HyperPoint.hpp:42]: Ultimately, we can't do the automatic conversion using aggregate initialization so we would need to provide a constructor (or force the user to provide the correct type).

> I really need to learn more about how variadic templates work. Could you please point me to the place where it explains how decay works for a parameter pack? I can't see anything like that [here](https://en.cppreference.com/w/cpp/language/parameter_pack).

I need `decay_t` here since the type might be deduced to `T&&`.

Comment (masterleinad) [reply src/geometry/ArborX_HyperPoint.hpp:42]:
```
/tmp/ArborX/test/tstQueryTreeCallbacks.cpp:102:23: error: non-constant-expression cannot be narrowed from type 'double' to 'float' in initializer list [-Wc++11-narrowing]
        points(i) = {{(double)i, (double)i, (double)i}};
                      ^~~~~~~~~
```
using https://github.com/aprokop/ArborX/pull/17 for example.

Comment (aprokop) [reply src/geometry/ArborX_HyperPoint.hpp:42]: I think that is fine. In the sense, that actually is something a user needs to address.

Comment (masterleinad) [reply src/geometry/ArborX_HyperPoint.hpp:42]: Yes, it's just that this pull request doesn't address all problems in aprokop#17.

**Comment by aprokop** [comment]
Other than removing `static_assert`, the PR looks good to me.

**Comment by masterleinad** [comment]
> Other than removing `static_assert`, the PR looks good to me.

Did that and also added a couple compile-time tests.

File: test/tstCompileOnlyAccessTraits.cpp [hunk 4]
```diff
@@ -82,6 +82,22 @@ void test_access_traits_compile_only()
...
+  static_assert(std::is_same_v<decltype(Point{1}), Point<1, float>>);
+  static_assert(std::is_same_v<decltype(Point{1.}), Point<1, double>>);
+  static_assert(std::is_same_v<decltype(Point{1., 2}), Point<2, double>>);
+  static_assert(std::is_same_v<decltype(Point{1, 2.}), Point<2, double>>);
+  static_assert(std::is_same_v<decltype(Point{2, 2}), Point<2, float>>);
+  static_assert(std::is_same_v<decltype(Point{2., 3.f, 2.}), Point<3, double>>);
+  static_assert(
+      std::is_same_v<decltype(Point{2.f, 3.f, 2.f}), Point<3, float>>);
+  static_assert(
+      std::is_same_v<decltype(Point<3, int>{2, 3, 2}), Point<3, int>>);
```
Comment (aprokop) [inline test/tstCompileOnlyAccessTraits.cpp:98]: Why did you need to specify the full type before the constructor here?

Comment (masterleinad) [reply test/tstCompileOnlyAccessTraits.cpp:98]: I just wanted to test/demonstrate that you can still construct a Point with a non-floating point type when specifying the template arguments explicitly. This would break with the `static_assert` I initially added.

**Review by aprokop — DISMISSED** [review]
LGTM

**Comment by masterleinad** [comment]
Retest this please.

**Comment by aprokop** [comment]
Couple issues.

First, seems like we would want to have at least one template argument:
```
arborx/src/geometry/ArborX_HyperPoint.hpp:34:25:
warning: zero size arrays are an extension [-Wzero-length-array]
  FloatingPoint _coords[DIM] = {};
                        ^~~
arborx/test/tstCompileOnlyAccessTraits.cpp:88:41: note:
in instantiation of template class
'ArborX::ExperimentalHyperGeometry::Point<0>' requested here
  static_assert(std::is_same_v<decltype(Point{}), Point<0, float>>);
```

Second, can we make `Point(a, b, c)` work in addition to `Point{a, b, c}`? I don't think it works at the moment.

**Comment by aprokop** [comment]
> If we want that, `ExperimentalHyperGeometry::Point` can be an aggregate anymore since we need to define constructors and we would probably end up with the same structure (mess?) as in `Point` if we want to support all the ways we construct `Point`.

That's fair. It also means that replacing `ArborX::Point` with a hyper-point will be a breaking change. Kinda expected it already. Guess if we want backward compatibility, we would need to provide a constructor for `Point<3, float>`, but then we would have to maintain that too.

**Comment by aprokop** [comment]
@masterleinad Please fix the formatting in the last patch.

**Comment by masterleinad** [comment]
> @masterleinad Please fix the formatting in the last patch.

Done.

**Review by aprokop — APPROVED** [review]

**Comment by aprokop** [comment]
Merging without HIP.
